Add Product Password Management for WooCommerce 8.1+ with Application Password Login

pmusolino opened the pull request:
Closes: #13387

## Description
This pull request implements the ability to manage product passwords when the user logs in using wp-admin credentials (application password). This feature is available in stores with WooCommerce version 8.1 or later, leveraging the new `password` field introduced in the WooCommerce REST API.

## Main Changes

- **Addition of the "password" field to product models:** A new `password` field has been added to the `EditableProductModel` and `EditableProductVariationModel` models.
- **Introduction of `ProductPasswordEligibilityUseCase`:** This new utility class checks if the WooCommerce store in use is eligible for the new product password management, which requires WooCommerce 8.1 or higher. The class verifies the installed and active WooCommerce plugin version, returning `true` if the version is 8.1.0 or higher, otherwise `false`.
- **Update of `ProductFormRemoteActionUseCase`:** This class has been modified to handle the product password update based on the WooCommerce version. For versions 8.1 and later, the password is updated locally in the `Product` model. For earlier versions, the update is performed via the `updateSitePostPassword` API.
- **User interface modifications:** Changes have been made to the user interface to allow users to set and modify product passwords. For instance, the cell for managing the product password is selectable only if the store is eligible for password modification.
- **Tests:** New unit tests have been added to verify the correct functioning of the new feature and the `ProductPasswordEligibilityUseCase` class.

## Functionality

1. The app checks if the user is authenticated without WPCom. If the user is authenticated with wp-admin credentials, the product password is retrieved from the `Product` model, provided the store uses WooCommerce 8.1 or higher.
2. Before making the product password field editable, the app checks if the store is eligible for the new password management via the `ProductPasswordEligibilityUseCase` class. A store is considered eligible if it uses WooCommerce 8.1 or higher, ensuring the presence of the new `password` field in the API.
3. If the store is eligible, the product password is managed directly in the `Product` model. Otherwise, the app uses the `updateSitePostPassword` API to update the password, and only if the user is logged through WP.com.


## Testing information
1. **Stores with WooCommerce 8.1 or higher and Application Password login:**
        - Verify that the app correctly retrieves the product password from the `Product` model when the user is authenticated with wp-admin credentials.
        - Verify that the app correctly updates the product password in the `Product` model when the user makes changes.
        - Verify that the app sends updates to the server only when necessary (e.g., when the password has actually been changed).

2. **Stores with WooCommerce versions lower than 8.1 and Application Password login:**
        - Verify that the app doesn't allow to select the visibility status.

3. **Stores with WooCommerce versions lower than 8.1 and WP.com login:**
        - Verify that the app correctly uses the `updateSitePostPassword` API to update the product password.
        - Verify that the app sends the password to the server only when necessary (e.g., when the password has actually been changed).

### Additional Testing
- Replicate the above conditions during product creation.
- Replicate the above conditions during product variation editing and creation.
- Given the changes to how the password works, and considering that the password field is connected to many other points in the product editing process, it would be beneficial to conduct a smoke test of the entire product.


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.
- [x] This PR includes refactoring; smoke testing of the entire section is needed.

pmusolino committed:
feat: Add ProductPasswordEligibilityUseCase for WooCommerce password visibility management

- Introduced `ProductPasswordEligibilityUseCase` class to determine eligibility for the new password endpoint in WooCommerce 8.1.
- Added `isEligibleForNewPasswordEndpoint()` method to check the WooCommerce plugin's active status and version.
- Implemented `getSystemPlugin(siteID:)` helper method to retrieve the WooCommerce plugin from stored plugins.
- Utilized `VersionHelpers.isVersionSupported` to ensure compatibility with the minimum required WooCommerce version.
- Integrated `StoresManager` and `StorageManagerType` for session and storage management.

pmusolino committed:
Merge commit '719a96491634710aa422b5eb86deaa167b072c12' into issue/13387-add-password-in-product-with-new-field-from-wc-8-1

pmusolino committed:
feat: implemented ProductPasswordEligibilityUseCaseTests

pmusolino committed:
fix: spaces

pmusolino committed:
fix: ProductPasswordEligibilityUseCaseTests

pmusolino committed:
feat: temp code to check if everyhting works properly with password in `ProductSettingsViewModel`

pmusolino committed:
Merge branch 'issue/13387-password-networking-tests' into issue/13387-add-password-in-product-with-new-field-from-wc-8-1

pmusolino committed:
Merge branch 'issue/13387-product-password-yosemite-layer' into issue/13387-add-password-in-product-with-new-field-from-wc-8-1

pmusolino committed:
feat: added password field in missing classes

pmusolino committed:
feat: product settings row now is tappable

pmusolino committed:
Merge commit 'd6c55aff02d1e572a441c994018c051267ae1476' into issue/13387-add-password-in-product-with-new-field-from-wc-8-1

pmusolino committed:
Merge commit 'b558bb82efa0f27f3b8b0101736d54a51442190b' into issue/13387-add-password-in-product-with-new-field-from-wc-8-1

pmusolino committed:
Merge commit '9b096e57a1d378de5863a91c2772d0095e6107a8' into issue/13387-add-password-in-product-with-new-field-from-wc-8-1

pmusolino committed:
Merge commit 'bf231fc715f221e13db715e946444309fca39a8b' into issue/13387-add-password-in-product-with-new-field-from-wc-8-1

pmusolino committed:
Merge commit '2a495769ebfd9b3507aadea5cb89434d5faee67f' into issue/13387-add-password-in-product-with-new-field-from-wc-8-1

pmusolino committed:
feat: updated logic for supporting password editing from WC 8.1 in `ProductFormRemoteActionUseCase`

pmusolino committed:
feat: edited existing tests and added new tests in `ProductFormRemoteActionUseCaseTests` for supporting password editing in WC above and below 8.1

pmusolino committed:
fix: trailing spaces

pmusolino committed:
feat: implemented logic in ProductSettingsViewModel to support password editing under WC 8.1

pmusolino committed:
feat: implemented unit tests in `ProductSettingsViewModelTests` to support password editing under WC 8.1

pmusolino committed:
fix: trailing spaces

pmusolino committed:
fix: insert sample system plugin into storage manager

pmusolino committed:
fix: space

pmusolino committed:
update: release notes to include password setup for Products with wp-admin credentials in WooCommerce 8.1+

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman commented:
Since this is a big change, I'm trying to test more thoroughly.

So far, for this case **Stores with WooCommerce 8.1 or higher and Application Password login** creating, updating, update behavior, removing, all work properly in my test.

hafizrahman commented:
These test cases also look good to me:
- **Stores with WooCommerce versions lower than 8.1 and Application Password login:**
- **Stores with WooCommerce versions lower than 8.1 and WP.com login**

hafizrahman commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Product Settings/ProductPasswordEligibilityUseCase.swift:
```diff
@@ -0,0 +1,55 @@
+    /// In WooCommerce 8.1, the REST API now includes a new field in the products response for managing visibility passwords.
+    /// This enables mobile apps to support this feature when accessing the site with Application Passwords.
+    /// https://github.com/woocommerce/woocommerce/pull/39438
+    func isEligibleForNewPasswordEndpoint() -> Bool {
```
np: the term "new" might not be accurate anymore in the future. It might not even be accurate now as well, since 8.1 was released almost a year ago. 

Perhaps `isEligibleForWooProductPasswordEndpoint` is more accurate? Or if it's too long, then `isEligibleForWooPasswordEndpoint`.

pmusolino replied:
Good suggestion, I updated it here e67596c25ebf025e55ad7ec603546adf74cd92b9

hafizrahman commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Product Settings/ProductSettingsRows.swift:
```diff
@@ -102,10 +102,12 @@ enum ProductSettingsRows {
-            /// If the password was not fetched for user authenticated with WPCom,
-            /// the cell is not selectable
-            if settings.password == nil && passwordProtectedAvailable {
+            /// Check if the store is eligible for password editing or is authenticated without WPCom.
```
Should this comment say "is authenticated with WPCom" instead?

And with that being said, wdyt about moving the `.isAuthenticatedWithoutWPCom == false` check to the first case just as a bit of optimization (remove the need to check db if auth with wpcom site).

pmusolino replied:
Good catch and good suggestion. I've made the change here: 6533335804d45842b3ced53c78e50d4d48a0f035

pmusolino committed:
Merge branch 'trunk' into issue/13387-add-password-in-product-with-new-field-from-wc-8-1

pmusolino committed:
refactor: the term "new" in a method might not be accurate anymore in the future. Perhaps `isEligibleForWooProductPasswordEndpoint` is more accurate

pmusolino committed:
Merge branch 'trunk' into issue/13387-add-password-in-product-with-new-field-from-wc-8-1

pmusolino committed:
Merge branch 'issue/13387-add-password-in-product-with-new-field-from-wc-8-1' of https://github.com/woocommerce/woocommerce-ios into issue/13387-add-password-in-product-with-new-field-from-wc-8-1

pmusolino committed:
Refactor: `ProductSettingsRows` for clarity and optimization

- Updated comment to correctly reflect the authentication condition.
- Optimized the conditional check by moving `.isAuthenticatedWithoutWPCom == false` to the first case to avoid unnecessary database checks when authenticated with WPCom.

staskus commented:
@pmusolino modifying the milestone to 20.2

pmusolino committed:
Merge commit '76c1467a28b13837d589a1680db4b464d18154b7' into issue/13387-add-password-in-product-with-new-field-from-wc-8-1

pmusolino committed:
fix: failing unit tests

pmusolino commented:
@selanthiraiyan @itsmeichigo When you have some time, could you please review this PR so I can merge it? 🙇‍♂️ Thanks!

itsmeichigo committed:
Merge branch 'trunk' into issue/13387-add-password-in-product-with-new-field-from-wc-8-1

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Product Settings/ProductPasswordEligibilityUseCase.swift:
```diff
@@ -0,0 +1,55 @@
+    private func getSystemPlugin(siteID: Int64) -> Yosemite.SystemPlugin? {
+        return storageManager.viewStorage
+            .loadSystemPlugins(siteID: siteID).map { $0.toReadOnly() }
+            .first(where: { $0.name == Constants.wcPluginName })
```
Even though the WooCommerce plugin name is unlikely to change, I feel like this is a fragile check. Usually we check for the plugin slug, but with the recent breaking change of the API, comparing the slug might be fragile as well. I think our [implementation](https://github.com/woocommerce/woocommerce-ios/blob/531676ced86f77c774d51b2fe5aa15f0bd316ffa/WooCommerce/Classes/Tools/CardPresentPluginsDataProvider.swift#L98) for card-present payments seem to be the most reliable. Should we follow the same method of checking plugins?

pmusolino replied:
Good suggestion. I updated the code here: 0f3503aaab6117ef296cb661a8d0c28c1c3651e7 now it should be more reliable.

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/ProductFormRemoteActionUseCaseTests.swift:
```diff
@@ -100,12 +106,81 @@ final class ProductFormRemoteActionUseCaseTests: XCTestCase {
+                                              active: true)
+        storageManager.insertSampleSystemPlugin(readOnlySystemPlugin: plugin)
+
+        let sut = ProductPasswordEligibilityUseCase(stores: storesManager, storageManager: storageManager)
```
I'm confused - this test should be added in a separate test file for `ProductPasswordEligibilityUseCase` instead of `ProductFormRemoteActionUseCase`.

pmusolino replied:
Updated here 31561985e4a6fa81f0fc08a879163acf13bb00b6 and here 8932bdff00a33fdafcce171b2443e69b2c08b571

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/ProductFormRemoteActionUseCaseTests.swift:
```diff
@@ -100,12 +106,81 @@ final class ProductFormRemoteActionUseCaseTests: XCTestCase {
+        let sut = ProductPasswordEligibilityUseCase(stores: storesManager, storageManager: storageManager)
+
+        // When
+        let isEligible: Bool = sut.isEligibleForWooProductPasswordEndpoint()
```
Same comment as above about the misplaced test.

pmusolino replied:
Updated with above commits 👍

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormRemoteActionUseCase.swift:
```diff
@@ -241,9 +247,14 @@ private extension ProductFormRemoteActionUseCase {
+        // Otherwise, update the password locally in the Product model, introduced in WC 8.1.
         guard let updatedPassword = password,
-              stores.isAuthenticatedWithoutWPCom == false else {
+              stores.isAuthenticatedWithoutWPCom == false || !ProductPasswordEligibilityUseCase().isEligibleForWooProductPasswordEndpoint() else {
```
The comment suggests that all conditions should be met. Should we replace `||` with `,` or `&&`?

pmusolino replied:
The comment here was not very clear, as it is correct to have an OR condition. Therefore, I updated the comment to make it clearer: 3bb6bc4c9331aa45e97bb6790775ee1e6982ba86

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Product Settings/ProductSettingsRows.swift:
```diff
@@ -102,10 +102,12 @@ enum ProductSettingsRows {
         }
 
         func handleTap(sourceViewController: UIViewController, onCompletion: @escaping (ProductSettings) -> Void) {
+            /// Check if the store is eligible for password editing or is authenticated with WPCom.
+            /// If neither condition is met, the cell will not be selectable.
```
I tested on Android and found that they allow visibility editing with 2 options public and private. Should we follow the same? If not, maybe it's better to hide the row entirely or display "Not editable" instead of "Public". The current experience makes it look like a bug, with the row tappable but no navigation.

pmusolino replied:
`Public` here is the "visibility" of the product, so I think it's better to display `Public` instead of `Not Editable`.
So, probably here the solution is something in the middle: display a small alert that explain to the user the status and why they cannot edit it. Wdyt?

itsmeichigo replied:
Yup, that sounds like a better solution. Please feel free to take care of it in a separate PR.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Product Settings/ProductSettingsViewModel.swift:
```diff
@@ -41,9 +41,20 @@ final class ProductSettingsViewModel {
+            }
+            /// If user is not on WC 8.1, and if the password is empty/nil,
+            /// and user is authenticated with WP.com, enter here, otherwise we skip this.
+            else if (product.password == nil || product.password?.isEmpty == true) && ServiceLocator.stores.isAuthenticatedWithoutWPCom == false {
```
Can you explain why we ignore the `password` field sent through the initializer and check only the field in the product?

pmusolino replied:
Probably this condition can be improved to make it clear, or even the comment can be improved. The reason is that `product.password` is available only if the user is on WC 8.1, otherwise the password will be fetched through Site post API.

itsmeichigo reviewed (commented):
I tested on simulator iPhone 15 iOS 17.4 with the 4 cases:
- Store with WC version > 8.1, authenticated with and without WPCom.
- Store with WC version < 8.1, authenticated with and without WPCom.
The PR works as described, but I have some questions in the comments.

pmusolino committed:
Merge commit 'dcfbf67c24fec624bd1935a465aeff373a0aaf01' into issue/13387-add-password-in-product-with-new-field-from-wc-8-1

pmusolino committed:
update: plugin name references
- Change `wcPluginName` from "WooCommerce" to "woocommerce" in `ProductPasswordEligibilityUseCase.swift`
- Update `pluginName` to "woocommerce" and replace occurrences in `ProductPasswordEligibilityUseCaseTests.swift`

pmusolino committed:
Merge branch 'issue/13387-add-password-in-product-with-new-field-from-wc-8-1' of https://github.com/woocommerce/woocommerce-ios into issue/13387-add-password-in-product-with-new-field-from-wc-8-1

pmusolino committed:
refactor: product password eligibility tests
- Add tests for WooCommerce version below 8.1 in `ProductPasswordEligibilityUseCaseTests.swift`.
- Add tests for WooCommerce version above 8.1 in `ProductPasswordEligibilityUseCaseTests.swift`.
- Update plugin initialization in `ProductFormRemoteActionUseCaseTests.swift`.

pmusolino committed:
Add tests for WooCommerce version eligibility in ProductPasswordEligibilityUseCaseTests
- Add test for eligibility when WooCommerce version is 8.1

pmusolino committed:
update: modified comment to clarify conditions for updating product password

pmusolino committed:
update: RELEASE-NOTES.txt. Add note about users setting up passwords on Products with WooCommerce 8.1 or higher

itsmeichigo committed:
Merge branch 'trunk' into issue/13387-add-password-in-product-with-new-field-from-wc-8-1

itsmeichigo reviewed (approved):
Thanks for the updates @pmusolino! Regarding the alert for the disabled update of visibility, please feel free to handle it in a separate PR.

pmusolino committed:
Merge branch 'trunk' into issue/13387-add-password-in-product-with-new-field-from-wc-8-1

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/222">20.3</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

pmusolino commented:
Thanks, @itsmeichigo. I opened a PR here to keep track of the alert issue and not forget about it https://github.com/woocommerce/woocommerce-ios/issues/13862
